keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+
+Contextual logging replaces the global logger by passing a `logr.Logger`
+instance into functions via a `context.Context` or an explicit
+parameter, building on top of [structured
+logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging).
+
+It enables the caller to:
+- attach key/value pairs that get included in all log messages
+- add names that describe which component or operation triggered a log messages
+- reduce the amount of log messages emitted by a callee by changing the
+  verbosity
+
+This works without having to pass any additional information into the callees
+because the additional information or settings are stored in the modified
+logger instance that is passed to it.
+
+Third-party components that use Kubernetes packages like client-go are no
+longer forced to use klog. They can choose an arbitrary implementation of
+`logr.Logger` and configure it as desired.
+
+During unit testing, each test case can use its own logger to ensure that log
+output is associated with the test case.
+
+## Motivation
+
+### Goals
+
+- Remove `k8s.io/klog` imports and usage from all packages.
+- Grant the caller of a function control over logging inside that function.
+- Provide documentation and helper code for setting up logging in unit tests.
+- Change as few exported APIs as possible.
+
+### Non-Goals
+
+- Remove the klog text output format.
+
+## Proposal
+
+Log calls that were already converted to structured logging need to be updated
+to use a logger that gets passed into the function in one of two ways:
+- as explicit parameter
+- attached to a `context.Context`
```

Comment:
Do we want to remove the global logger from the entire code?
Or do we want to add a way to do contextual logging as an option?

Reply:
Contextual logging will be optional at first and disabled (see feature gate discussion). The global logger in klog will continue to be used during the transition period. Later the global klog logger will be replaced with a global logger in `k8s.io/klogr` which acts as fallback when code doesn't want to or can't use contextual logging.